contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -3851,16 +3871,196 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+					{
+						Node *func = (Node *) linitial(ob->objname);
+						funcname = strVal(func);
+						logicalschema = get_authid_user_ext_schema_name(dbname, current_user);
+					}
+					else
+					{
+						Node *schema = (Node *) linitial(ob->objname);
+						char *schemaname = strVal(schema);
+						Node *func = (Node *) lsecond(ob->objname);
+						logicalschema = get_logical_schema_name(schemaname, true);
+						funcname = strVal(func);
+					}
+
+					/* If ALL PRIVILEGES is granted/revoked. */
+					if (list_length(grant->privileges) == 0)
+					{
+						/*
+						 * Case: When ALL PRIVILEGES is revoked internally during create function.
+						 * pstmt->stmt_len = 0 means it is an implicit REVOKE statement issued at the time of create function/procedure.
+						 * For more details, please refer revoke_func_permission_from_public().
+						 * If schema entry exists in the catalog, implicitly grant permission on the new object to the user.
+						 */
+						if ((pstmt->stmt_len == 0) && privilege_exists_in_bbf_schema_permissions(logicalschema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, NULL, true))
+						{
+							call_prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params, queryEnv, dest, qc);
+							exec_internal_grant_on_function(logicalschema, funcname, obj_type);
+							return;
+						}
+						
+						foreach(lc, grant->grantees)
+						{
+							RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+							if (grant->is_grant)
+							{
+								add_or_update_object_in_bbf_schema(logicalschema, funcname, ALL_PERMISSIONS_ON_FUNCTION, rol_spec->rolename, obj_type, true, func_args);
+							}
+							else
+							{
+								/*
+								 * 1. If permission on schema exists, don't revoke any permission from the object.
+								 * 2. If permission on object exists, update the privilege in the catalog and revoke permission.
+								 */
+								update_privileges_of_object(logicalschema, funcname, ALL_PERMISSIONS_ON_FUNCTION,  rol_spec->rolename, obj_type, false);
+								if (privilege_exists_in_bbf_schema_permissions(logicalschema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, rol_spec->rolename, false))
+									return;
+							}
+						}
+						call_prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params, queryEnv, dest, qc);
+					}
+					foreach(lc1, grant->privileges)
+					{
+						AccessPriv *ap = (AccessPriv *) lfirst(lc1);
+						AclMode privilege = string_to_privilege(ap->priv_name);
+						foreach(lc, grant->grantees)
+						{
+							RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+							if (grant->is_grant)
+							{
+								call_prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params, queryEnv, dest, qc);
+								/*
+								 * If it is an implicit GRANT issued by exec_internal_grant_on_function, then we should not add catalog
+								 * entry. Catalog entry is supposed to be added only by explicit GRANTs.
+								 */
+								if (strcmp("(GRANT STATEMENT )", queryString) != 0)
```

Comment:
We should move this outside the loop as well. if (strcmp("(GRANT STATEMENT )", queryString) == 0), is_grant = true and in that case, we're not doing anything. 

Do we need to consider the case for ALL PRIVILEGE case as well?

Reply:
if (strcmp("(GRANT STATEMENT )", queryString) == 0), is_grant = true  -> We are executing the pg command by process utility but not updating the catalog.

`Do we need to consider the case for ALL PRIVILEGE case as well?` No we don't have to. There is no such case.

Reply:
Again, this is a loop-invariant. Skip the loop if strcmp("(GRANT STATEMENT )", queryString)  == 0.